spyder_kernels/customize/spydercustomize.py

```diff
@@ -671,9 +676,48 @@ def runfile(filename, args=None, wdir=None, namespace=None, post_mortem=False):
     sys.argv = ['']
     namespace.pop('__file__')
 
+
 builtins.runfile = runfile
 
 
+def runcell(cellname, filename):
+    """
+    Run a code cell from an editor as a file.
+
+    Currently looks for code in in a `ipython` property called `cell_code`.
+    This property must be set by the editor prior to calling this function.
+    This function deletes the contents of `cell_code` upon completion.
+
+    Parameters
+    ----------
+    cellname : str
+        Used as a reference in the history log of which
+        cell was run with the fuction. This variable is not used.
+    filename : str
+        Needed to allow for proper traceback links.
+    """
+    try:
+        filename = filename.decode('utf-8')
+    except (UnicodeError, TypeError, AttributeError):
+        # UnicodeError, TypeError --> eventually raised in Python 2
+        # AttributeError --> systematically raised in Python 3
+        pass
+    run_umr()
+    ipython_shell = get_ipython()
+    try:
+        cell_code = ipython_shell.cell_code
+    except AttributeError:
+        _print("--Run Cell Error--\n"
+               "Please use only through Spyder's Editor; "
+               "shouldn't be called manually from the console")
+        return
+    ipython_shell.run_cell(cell_code)
+    del ipython_shell.cell_code
```

Comment:
>So you won't be able to rerun a previous cell ran with runcell? I think that would seem a bit limiting to our users...

This is unfortunate but I think this would be a 'ah ok' moment for most users and not a deal breaker, since users can't run a cell from the console right now. Cell based work is mostly all in the editor anyway (change code -> hotkey) it seems unlikely that a user would return to the console after making a change to a cell in the editor.

I guess the ideal interface is call `runcell('my_cell', 'my_file')` and have `runcell` execute the right code cell in the right file. But code cells without names or the same name end that idea. 

Another interface is just call `runcell()` in the console and the IPython kernel asks the editor for the code in the cell that the cursor is in, but I believe this isn't currently possible with qtconsole and `runcell()` still might not do what users expect

All these things considered it seems code cells are best run from the code cell itself and not the console. This mirrors the notebook experience.


>So for that to work (from the UX point of view) we probably need to avoid saving runcell in IPython's history.

I agree, it certainly shouldn't be in the "up button" for previously used commands.  It's not useful there at all.  I'll see what I can do, but the difficulty of changing this command history was what lead me to make `runcell`.

My original goal of all this was to have a filter on the history to not include code executed from code cells and `runfile` ( see https://github.com/spyder-ide/spyder/issues/7233) but there was no way to differentiate code cells from anything else because they where just copied code just like anything else.  I wanted this because pressing up and down through long blocks of code was pretty error prone and the history log wasn't overly useful.

At this point it only has a name and a file as the arguments because I thought it would be good have an indication to the user of what code it was running currently. Having this function call in the history log might help the user see what code was run at that point to give context to the surrounding commands

Reply:
One more thing runfile also doesn't save my file when I run it from the console which would not be expected by the user.

TLDR;

I think that history filtering would be great and is best left for another PR because far as I remember I think we would have make changes to qtconsole to do it properly.

Reply:
Ok, I can help you with that, don't worry. I think I know how to do it without much trouble (probably from the Spyder side, not here).

For now, please fix the little grammar errors me and @CAM-Gerlach pointed out in one of your docstrings, then I'll merge this one.

Great work, by the way!! I really like your solution!

Reply:
Also, do you need help with a test? I can help you with that too.

Reply:
>For now, please fix the little grammar errors me and @CAM-Gerlach pointed out in one of your docstrings, then I'll merge this one.

I'll fix this up tonight.

>Great work, by the way!! I really like your solution!

It work pretty well now. Lots of ideas from @jitseniesen in here too!